### ISSUE TITLE
First AM message is always received as eager

### ISSUE THREAD

**Issue by pentschev** [openucx/ucx#6721]
When using the Active Message API in UCP with UCX master, the first message is always received as eager, even if it's larger than `UCX_RNDV_THRESH`.

After discussing this with @brminich , it became clear that this only happens for the first message exchanged between an EP pair, whereas subsequent messages -- after wireup completse -- respect rendezvous. It would be nice to have this documented if it's not possible to deal with this issue in some other way.

**Comment by yosefe** [comment]
@pentschev the AM API does not guarantee that any specific protocol would be used, unless EAGER/RNDV send flags are passed to the active message send function.
A portable application should not rely on a particular value of UCX_RNDV_THRESH, which is indented for runtime tuning purposes.

**Comment by pentschev** [comment]
> @pentschev the AM API does not guarantee that any specific protocol would be used, unless EAGER/RNDV send flags are passed to the active message send function.

How come? Shouldn't it respect the `UCX_RNDV_THRESH`? In particular, we are interested in making sure that RNDV is used for transfers above that size to ensure CUDA IPC is used when available.

> A portable application should not rely on a particular value of UCX_RNDV_THRESH, which is indented for runtime tuning purposes.

That makes sense, but in this case where I encountered this was when I was testing specifically for the RNDV behavior to ensure CUDA IPC is used when available.

**Comment by shamisp** [comment]
IMHO, there should be a better way to extract this info. Maybe some statistic about CUDA IPC usage.

**Comment by pentschev** [comment]
It's not only about extracting that information, in the UCX-Py case above where I first encountered this I was deliberately trying to test the rendezvous codepath with CUDA transfers. I could also set the RNDV flag to do that, but that wouldn't guarantee I'm testing the real condition, which today is not setting either EAGER or RNDV flags but letting UCX do that.

**Comment by yosefe** [comment]
@pentschev what happens if you do not set RNDV threshold? worse performance?
I would recommend to check the recv_attr in UCX-Py AM callback, regardless of UCX_RNDV_THRESH tuning:
`recv_attr & UCP_AM_RECV_ATTR_FLAG_RNDV`
https://github.com/openucx/ucx/blob/a08beda2d112cbdf52f6d57e0b11c8bb49e2f4bd/src/ucp/api/ucp.h#L1627

**Comment by pentschev** [comment]
> @pentschev what happens if you do not set RNDV threshold? worse performance?

Once wireup completes, whatever RNDV threshold is set (including the `8K` default) is respected, leading to what I expect. Before wireup, any message will be an EAGER transfer, implying a host copy, and then leading to worse performance.

> I would recommend to check the recv_attr in UCX-Py AM callback, regardless of UCX_RNDV_THRESH tuning:
> `recv_attr & UCP_AM_RECV_ATTR_FLAG_RNDV`

We do that in the same way that the iodemo does: https://github.com/openucx/ucx/blob/01b4d70b4f01c539dee0a87b03d150b34da50fca/test/apps/iodemo/ucx_wrapper.cc#L761-L781 

The above works as expected but only after wireup completes, which is what led me to believe something was wrong but it's only because of wireup. To be clear, I don't think this is a bug, but perhaps something that could be documented as I wasn't really sure what was happening until @brminich told me this was probably due to wireup, which I then confirmed by adding a small transfer preceding the CUDA data transfer I was expecting to respect the RNDV threshold.

**Comment by pentschev** [comment]
I was working on support AM in UCX-Py without a listener (today the only way to establish connections using AM), by allowing creating an endpoint directly to a remote worker via its address, but the behavior is different from the above.

When a listener is involved, after sending a wireup message, messages larger than `UCX_RNDV_THRESH` will be received via rndv. By looking at logs with `UCX_LOG_LEVEL=DATA`, I see that `rndv_thresh` is indeed agreed upon by both parties after wireup:

```
[1633093429.115101] [prm-dgx-04:26776:0]          ucp_am.c:886  UCX  REQ   select am request(0x55d551637940) progress algorithm datatype=0x8 buffer=0x7f6f668bcf90 length=6 header_length=4 max_short=-1 rndv_thresh=18446744073709551615 zcopy_thresh=18446744073709551615
[1633093429.174415] [prm-dgx-04:26673:0]          ucp_am.c:886  UCX  REQ   select am request(0x558f7af80f80) progress algorithm datatype=0x8 buffer=0x7f1756b9b790 length=6 header_length=4 max_short=-1 rndv_thresh=8192 zcopy_thresh=2153
[1633093429.505655] [prm-dgx-04:26673:0]          ucp_am.c:886  UCX  REQ   select am request(0x558f7af80e40) progress algorithm datatype=0x8 buffer=0x7f16ee000000 length=134217728 header_length=4 max_short=-1 rndv_thresh=8192 zcopy_thresh=1
[1633093429.175564] [prm-dgx-04:26776:0]          ucp_am.c:886  UCX  REQ   select am request(0x55d551637800) progress algorithm datatype=0x8 buffer=0x7f6c12000000 length=134217728 header_length=4 max_short=-1 rndv_thresh=8192 zcopy_thresh=1
```

However, when I have no listener, the `rndv_thresh` is never agreed to match that of `UCX_RNDV_THRESH` and all messages are received as eager:

```
[1633093446.699091] [prm-dgx-04:27052:0]          ucp_am.c:886  UCX  REQ   select am request(0x56470eb69b40) progress algorithm datatype=0x8 buffer=0x7f7b9d089070 length=6 header_length=4 max_short=-1 rndv_thresh=18446744073709551615 zcopy_thresh=2153
[1633093446.701932] [prm-dgx-04:26946:0]          ucp_am.c:886  UCX  REQ   select am request(0x55b17e9ef3c0) progress algorithm datatype=0x8 buffer=0x7fe55180e790 length=6 header_length=4 max_short=-1 rndv_thresh=18446744073709551615 zcopy_thresh=2153
[1633093446.702459] [prm-dgx-04:27052:0]          ucp_am.c:886  UCX  REQ   select am request(0x56470eb69b40) progress algorithm datatype=0x8 buffer=0x7f7842000000 length=134217728 header_length=4 max_short=-1 rndv_thresh=18446744073709551615 zcopy_thresh=1
[1633093446.989185] [prm-dgx-04:26946:0]          ucp_am.c:886  UCX  REQ   select am request(0x55b17e9ef3c0) progress algorithm datatype=0x8 buffer=0x7fe4fcdfa010 length=134217728 header_length=4 max_short=-1 rndv_thresh=18446744073709551615 zcopy_thresh=2153
```

I know @yosefe said we shouldn't rely on `UCX_RNDV_THRESH` with the AM API, but this also prevents us from testing the correct codepath. Plus, in the case I'm testing the message size is 128MB which is definitely going to perform better when using rndv as workers can communicate via CUDA IPC.

Therefore I have a few questions:

1. Is the behavior I'm describing a bug or is it by design?
2. How can we enforce `UCX_RNDV_THRESH` is respected so I can deterministically test the codepath?
3. What should we do in cases such as mentioned above in this comment, should we decide based on message size whether it should be sent via eager/rndv (i.e., in my understanding, we would transfer the decision from UCX back to the user application) or is UCX still supposed to decide based on HW available, etc.?

**Comment by yosefe** [comment]
it's expected that before wireup is complete, messages with RNDV would be sent as eager since at that point it's not yet known if rdma is possible. It should not impact overall performance since it's just the first message.
To ensure RDNV flow is tested, cand send small active message and do the large RNDV after it arrived.
if no EAGER/RDNV flag is passed, UCX will decide the threshold internally.

**Comment by pentschev** [comment]
> To ensure RDNV flow is tested, cand send small active message and do the large RNDV after it arrived.

That works, but only if I have a listener, without the listener UCX doesn't seem to respect it at all.

> if no EAGER/RDNV flag is passed, UCX will decide the threshold internally.

How can I verify what should be the threshold decided internally? As per the second piece of log above it seems to be defined as `rndv_thresh=18446744073709551615`, which effectively means rndv will never be used.

To add one more data point, I would just like to mention that I hacked `ucp_am_rndv_thresh` to test my hypothesis and always return `8192u`, and in that case messages larger than `8K` were indeed received via rndv, even without a listener.

**Comment by brminich** [comment]
do you mean you have to have a listener on client side as well for rndv to be selected?

**Comment by pentschev** [comment]
No, I have two setups, the first one (where rndv works) is as follows:

- Server:
  - Start worker and install AM callback;
  - Start listener with `ucp_listener_create`;
  - Create endpoint from `conn_request` and send on this endpoint, but receives directly on the worker and received messages can be mapped to this particular endpoint via `reply_ep`;
- Client:
  - Create endpoint to listener using it's IP/port and send/recv on this endpoint.

The second one use no listener but rndv is never triggered:

- Server:
  - Start worker and install AM callback;
  - Has no endpoint to client until a message is received and then use `reply_ep` to send messages;
- Client:
  - Creates endpoint to remote worker using the worker address only and send/recv on this endpoint.

Neither client process has a listener.

**Comment by brminich** [comment]
in the second case, is it server who always use eager or client? or both?

**Comment by pentschev** [comment]
Both of them always use eager in that case.

**Comment by brminich** [comment]
can you please collect traces with debug UCX version and UCX_LOG_LEVEL=data and share it somehow?

**Comment by pentschev** [comment]
Sure @brminich , I was doing exactly that when you typed. You'll find both traces in 
[am_logs.tar.gz](https://github.com/openucx/ucx/files/7268052/am_logs.tar.gz). Sorry I had to compress the file, one of them was too large for GH.

**Comment by brminich** [comment]
The problem is that rndv protocol is completely disabled when error handling is requested, but connection is not established in cm mode.
@evgeny-leksikov, are we going to get rid of this limitation?
https://github.com/openucx/ucx/blob/master/src/ucp/core/ucp_ep.c#L2862

**Comment by evgeny-leksikov** [comment]
@brminich sure, we have to add "keep alive" check to this condition

**Comment by yosefe** [comment]
@evgeny-leksikov @dmitrygx  isn't this "indirect id" condition which causes UCP EP to go over outstanding requests and purge them during ep close?

**Comment by evgeny-leksikov** [comment]
IIRC, we disable RNDV in case of err handling mode peer to make all requests completed in finite time. On other hand, if err is detected, user can close the failed EP and all outstanding requests will be completed during EP close. IMO this behavior is not well defined/documented.
